keps/6757-failure-recovery/README.md

```diff
@@ -0,0 +1,336 @@
+  FailureRecoveryPolicy *FailureRecoveryPolicy `json:failureRecoveryPolicy,omitempty"`
+}
+
+type FailureRecoveryPolicy struct {
+  // Rules specifies the rules to be enabled for failure recovery.  
+  Rules []FailureRecoveryRule `json:"rules"`
+}
+
+type FailureRecoveryRule struct {
+  // Action specifies the action taken to recover from failures.  
+  Action FailureRecoveryAction `json:"action"`
+
+  // PodLabelSelector specifies the scope of resources covered by failure recovery -
+  // resources not matching the selector behave according to the default Kubernetes mechanism.
+  PodLabelSelector *metav1.LabelSelector `json:"podLabelSelector,omitempty"`
+	
+  // PodTerminationConfig contains configuration for the `PodTerminationConfig` strategy.
+	// +optional
+  PodTerminationConfig *PodTerminationConfig `json:"podTerminationConfig,omitempty"`
+}
+
+type PodTerminationConfig struct {
+	// PodTerminationGracePeriod is the duration between when the pod's `deletionGracePeriodSeconds`
+  // elapses and when the pod should be forcefully deleted.
+  ForcefulTerminationGracePeriod *time.Duration `json:"forcefulTerminationGracePeriod"`
+}
+
+type FailureRecoveryAction string
+
+const (
+	PodTermination FailureRecoveryAction = "PodTermination"
+)
+```
+
+This API requires opt-in both from the administrators and user's end, as the workloads would
+have to be labeled in order for them to be covered with failure recovery.
+
+An example failure recovery configuration:
+
+```yaml
+failureRecoveryPolicy:
+  rules:
+  - action: PodTermination
+    podLabelSelector:
+      safe-to-fail: true
+    podTerminationConfig:
+      forcefulTerminationGracePeriod: 5m
+```
+
+### Default Grace Period
+
+The proposal is to **not** set a default value for the forceful termination grace period.
+This puts configuration over convention, as it will make the user consciously think about the
+grace period the need for their specific use case, without risking a faulty default.
+
+Since enabling the feature is inherently risky, this serves as an additional explicit
+acknowledgement of this risk from the user's end.
+
+Alternatively, a default value can be inferred from the `terminationGracePeriodSeconds`.
+For example, it could be a multiple of that value. Nevertheless, this still runs the risk of
+setting a very small value or even 0, depending on the user's configuration.
+
+### Controller
+
+Setting the strategy type to `PodTermination` will enable a controller, which manages
+the failed nodes.
+
+The controller has to ignore any updates to pods that:
+1. Are not terminating.
+    * `pod.DeletionTimestamp == nil || pod.DeletionGracePeriodSeconds == nil`
+1. Are in a terminal phase.
+    * `pod.Status.Phase != corev1.PodRunning && pod.Status.Phase != corev1.PodPending`
+1. Are not managed by Kueue.
+    * The `kueue.x-k8s.io/managed` label is set for pods with pod integration enabled.
+    * The presence of the `kueue.x-k8s.io/podset` label could be used in other cases.
```

Comment:
```suggestion
1. Are not managed by Kueue.
    * That means, have neither of the labels `kueue.x-k8s.io/managed` and `kueue.x-k8s.io/podset` set.
```

This would be clearer to me - especially given your convention from the bullets above.
(Or, if it sounds less clear to you, please at least tell me if I got it right :)